src/matcouply/penalties.py

```diff
@@ -1394,12 +1394,10 @@ def factor_matrices_update(self, factor_matrices, feasibility_penalties, auxes):
 
         # Construct matrix A to peform thomas algorithm on
 
-        K = len(B_is[0])
-
         A = (
-            np.diag([self._get_laplace_coef(i, K, self.smoothness_l) + rhos[i] for i, rho in enumerate(rhos)], k=0)
-            - np.diag(np.ones(K - 1) * 2 * self.smoothness_l, k=1)
-            - np.diag(np.ones(K - 1) * 2 * self.smoothness_l, k=-1)
+            np.diag([self._get_laplace_coef(i, I) + rhos[i] for i, rho in enumerate(rhos)], k=0)
+            - np.diag(np.ones(I - 1) * 2 * self.smoothness_l, k=1)
+            - np.diag(np.ones(I - 1) * 2 * self.smoothness_l, k=-1)
```

Comment:
If we use `tl.diag`, we'll support PyTorch as well :)

```suggestion
            tl.diag([self._get_laplace_coef(i, I) + rhos[i] for i, rho in enumerate(rhos)], k=0)
            - tl.diag(np.ones(I - 1) * 2 * self.smoothness_l, k=1)
            - tl.diag(np.ones(I - 1) * 2 * self.smoothness_l, k=-1)
```